### PR TITLE
WIP add types to part of store

### PR DESCRIPTION
**Pull request by mateuszkrzak**

### CODE REVIEW

File: app/src/containers/SearchPage/reducer.js [hunk 1]
```diff
@@ -8,7 +10,31 @@ import {
 } from './action-types';
 import createReducer from '../../redux/create-reducer';
 
-const initialState = {
+export type Movie = {
```
Comment (olvnikon) [inline app/src/containers/SearchPage/reducer.js]: If you are sure you never mutate the object you can use `+` to make fields readonly. And also don't forget about exact typing with `|`.
```javascript
export type Movie = {|
  +id: number,
  +title: string,
  ...
  +genres: Array<string>,
|};
```

File: app/src/containers/SearchPage/reducer.js [hunk 2]
```diff
@@ -8,7 +10,31 @@ import {
...
+  release_date: Date,
+  poster_path: string,
+  overview: string,
+  budget: number,
+  revenue: number,
+  runtime: number,
+  genres: Array<string>,
+}
+
+type SearchState = {
```
Comment (olvnikon) [inline app/src/containers/SearchPage/reducer.js]: The same for exact typing and readonly notation.

File: app/src/containers/SearchPage/actions.js [hunk 3]
```diff
@@ -1,29 +1,36 @@
...
-  SET_MOVIES_QUERY,
-  SET_MOVIES_SORT_BY,
-  SET_MOVIES_SEARCH_BY,
-} from './action-types';
+// @flow
+
 import moviesService from '../../services/MoviesService';
+import type { Movie } from './reducer';
+
+type FetchMoviesRequestAction = { type: "FETCH_MOVIES_REQUEST"};
```
[the remark this answers is not in the document]
Comment (mateuszkrzak) [reply app/src/containers/SearchPage/actions.js]: Thanks to my vcs plugin I can see what exactly is happening and remark from above gave me:
`type FetchMoviesRequestAction = { | +type: any |}`

I can explicitly type action-types to string, but it seems trivial and not really helpful.

File: .flowconfig [hunk 4]
```diff
@@ -0,0 +1,11 @@
+[ignore]
+.*/node_modules
```
[the remark this answers is not in the document]
Comment (mateuszkrzak) [reply .flowconfig]: All right.

File: app/src/containers/SearchPage/action-types.js [hunk 5]
```diff
@@ -1,6 +1,6 @@
-export const FETCH_MOVIES_REQUEST = 'FETCH_MOVIES_REQUEST';
+export const FETCH_MOVIES_REQUEST = "FETCH_MOVIES_REQUEST";
```
[the remark this answers is not in the document]
Comment (mateuszkrzak) [reply app/src/containers/SearchPage/action-types.js]: Ooops, this should not be commited.

File: app/src/containers/SearchPage/actions.js [hunk 6]
```diff
@@ -1,29 +1,36 @@
...
+type FetchMoviesFailureAction = { type: "FETCH_MOVIES_FAILURE", payload: string };
+type FetchMoviesSuccessAction = { type: "FETCH_MOVIES_SUCCESS", payload: { movies: Array<Movie>, moviesCount: number} };
```
[the remark this answers is not in the document]
Comment (mateuszkrzak) [reply app/src/containers/SearchPage/actions.js]: Keys are different, it's not the same.

File: app/src/containers/SearchPage/actions.js [hunk 7]
```diff
@@ -1,30 +1,65 @@
...
+|};
+export type SetMoviesSearchByAction = {|
+  +type: typeof SET_MOVIES_SEARCH_BY, 
+  +payload: string 
+|};
+
+function getMoviesRequest(): FetchMoviesRequestAction {
   return {
-    type: FETCH_MOVIES_REQUEST,
+    type: FETCH_MOVIES_FAILURE,
```
Comment (olvnikon) [inline app/src/containers/SearchPage/actions.js]: Shouldn't it be 'FETCH_MOVIES_REQUEST'?

File: app/src/containers/SearchPage/actions.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (olvnikon) [reply app/src/containers/SearchPage/actions.js]: Sure. Sorry, I forgot to add one essential comment. You should specify the type of the constant as well. Take a look:
```javascript
// Specify constant types
const FETCH_MOVIES_REQUEST: 'FETCH_MOVIES_REQUEST' = 'FETCH_MOVIES_REQUEST';
const FETCH_MOVIES_FAILURE: 'FETCH_MOVIES_FAILURE' = 'FETCH_MOVIES_FAILURE';

type FetchMoviesRequestAction = {| 
  +type: typeof FETCH_MOVIES_REQUEST 
|};

function getMoviesRequest_correct(): FetchMoviesRequestAction {
   return {
    type: FETCH_MOVIES_REQUEST, // Correct value
   };
 }

function getMoviesRequest_incorrect(): FetchMoviesRequestAction {
   return {
    type: FETCH_MOVIES_FAILURE, // Incorrect value
   };
 }
```

https://flow.org/try/#0PTAEGUAcFMGMEsBmBPUsD2A7AzgFwIaa6i7IzYBQGOxAYgKIAqAwgBID6AsgPIBqAkvXDsASvQCKAVSGMAXKADkDFhx4ChoidPCMFoALyLlbLn0HCxUmQoDcVLHlDHVZjbQCC-ADKSx8pUwmaubsHt6+9HqGASqm6sJhPmK2FBSkME7QuLAAFpzoAG7w0Ngi0ACOAK4luO6wuPBYBqAA3gA+oBSgoADU6dDy-eiIToEu8ZpWOp1tAL52FIiVmPWNmKAA5ln5RSVlVTXsGABOx3C4ABQAlPK0Wbk7xaUV1Xh1DU0tXd1nuJXH6y+3W6-VuYziIUs2kYABpQCBQMx0KdzqACvgADbVb6geZdWapJYrD7rLa4R57F6HeArZFnerXW73PKFJ77V61VafHG-f6AnEkMgDUaxYJuTxJehwhH8Wko+pozHY4F43EUIA

File: .flowconfig [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+.*/node_modules/cypress/*
+.*/node_modules/config-chain/*
+[include]
+
+[libs]
+
+[lints]
+
+[options]
+module.file_ext=.scss
```
Comment (olvnikon) [inline .flowconfig]: Unfortunately, this line overwrites the whole config of flow. It should be something like this:
```
module.use_strict=true
module.file_ext=.css
module.file_ext=.scss
module.file_ext=.js
module.file_ext=.jsx
module.file_ext=.json
```

File: app/src/containers/SearchPage/action-types.js [hunk 9]
```diff
@@ -3,4 +3,4 @@ export const FETCH_MOVIES_FAILURE = 'FETCH_MOVIES_FAILURE';
 export const FETCH_MOVIES_SUCCESS = 'FETCH_MOVIES_SUCCESS';
```
Comment (olvnikon) [inline app/src/containers/SearchPage/action-types.js]:
```javascript
export const FETCH_MOVIES_REQUEST: 'FETCH_MOVIES_REQUEST' = 'FETCH_MOVIES_REQUEST';
export const FETCH_MOVIES_FAILURE: 'FETCH_MOVIES_FAILURE' = 'FETCH_MOVIES_FAILURE';
export const FETCH_MOVIES_SUCCESS: 'FETCH_MOVIES_SUCCESS' = 'FETCH_MOVIES_SUCCESS';
export const SET_MOVIES_QUERY: 'SET_MOVIES_QUERY' = 'SET_MOVIES_QUERY';
export const SET_MOVIES_SORT_BY: 'SET_MOVIES_SORT_BY' = 'SET_MOVIES_SORT_BY';
export const SET_MOVIES_SEARCH_BY: 'SET_MOVIES_SEARCH_BY' = 'SET_MOVIES_SEARCH_BY';
```

File: app/src/containers/SearchPage/actions.js [hunk 10]
```diff
@@ -1,30 +1,65 @@
...
+  +payload: string 
+|};
+type FetchMoviesSuccessAction = {| 
+  +type: typeof FETCH_MOVIES_SUCCESS, 
+  +payload: { 
+    +movies: Array<Movie>,
+    +moviesCount: number
+  } 
+|};
+export type SetMoviesQueryAction = {|
```
Comment (olvnikon) [inline app/src/containers/SearchPage/actions.js]: You will not need to export these types:
```javascript
type SetMoviesQueryAction = {|
  +type: typeof SET_MOVIES_QUERY, 
  +payload: string 
|};
type SetMoviesSortByAction = {|
  +type: typeof SET_MOVIES_SORT_BY, 
  +payload: string 
|};
type SetMoviesSearchByAction = {|
  +type: typeof SET_MOVIES_SEARCH_BY, 
  +payload: string 
|};
```

File: app/src/containers/SearchPage/actions.js (hunk 7 above)
Comment (mateuszkrzak) [reply app/src/containers/SearchPage/actions.js]: You are right, thanks.

File: .flowconfig (hunk 8 above)
Comment (mateuszkrzak) [reply .flowconfig]: Thanks!

File: app/src/containers/SearchPage/action-types.js (hunk 9 above)
Comment (mateuszkrzak) [reply app/src/containers/SearchPage/action-types.js]: No luck with this, from vcs plugin I got:
`FetchMoviesRequestAction: type FetchMoviesRequestAction = { | +type: any |}`

File: app/src/containers/SearchPage/actions.js (hunk 10 above)
Comment (olvnikon) [reply app/src/containers/SearchPage/actions.js]: Because you will use "typeof" instead.

Comment (mateuszkrzak) [reply app/src/containers/SearchPage/actions.js]: Neat! Thanks.

**Pull request merged**